File: ios/StilldoShared/WidgetStore.swift
//
//  The one place the app and its widgets meet.
//
//  A widget runs in its own process and cannot open the app's SQLite file, so
//  the app writes a small summary into a shared container and the widgets read
//  only that. Both targets compile this file, so the shape can never disagree
//  between the side that writes it and the side that reads it.
//

import Foundation

/// Keep in step with `src/widgets/snapshot.ts`. The app group has to be on the
/// entitlements of both targets, and its name follows the bundle identifier —
/// changing one means changing the other.
enum WidgetStore {
  static let appGroup = "group.org.reactjs.native.example.StillDo"
  static let snapshotKey = "stilldo.snapshot"

  private static var shared: UserDefaults? {
    UserDefaults(suiteName: appGroup)
  }

  static func write(json: String) {
    shared?.set(json, forKey: snapshotKey)
  }

  /// Nil when nothing has been written yet — a widget added before the app has
  /// ever been opened — or when the payload was written by a build of the app
  /// whose shape this one does not know.
  static func read() -> WidgetSnapshot? {
    guard
      let json = shared?.string(forKey: snapshotKey),
      let data = json.data(using: .utf8),
      let snapshot = try? JSONDecoder().decode(WidgetSnapshot.self, from: data),
      snapshot.version == WidgetSnapshot.supportedVersion
    else { return nil }
    return snapshot
  }
}

struct WidgetTask: Codable, Hashable {
  let id: Int
  let title: String
  let note: String
}

struct WidgetSnapshot: Codable, Hashable {
  static let supportedVersion = 1

  let version: Int
  let rightNow: WidgetTask?
  let openCount: Int
  let caughtToday: Int
  let caughtDay: String
  let sweepTime: String
  let appearance: String

  /// What a widget shows before it has ever been given anything: the gallery
  /// preview, and the first seconds of a widget added before the app is opened.
  static let empty = WidgetSnapshot(
    version: supportedVersion,
    rightNow: nil,
    openCount: 0,
    caughtToday: 0,
    caughtDay: "",
    sweepTime: "21:00",
    appearance: "system"
  )

  /// The tally is about the day it was written on. A widget outlives that, so
  /// after midnight it reports nothing caught rather than yesterday's count.
  func caughtToday(on day: String) -> Int {
    caughtDay == day ? caughtToday : 0
  }
}

/// The URLs the widgets open the app with; parsed by `src/widgets/links.ts`.
enum WidgetLink {
  static let voice = URL(string: "stilldo://capture/voice")!
  static let text = URL(string: "stilldo://capture/text")!
  static let photo = URL(string: "stilldo://capture/photo")!
  static let now = URL(string: "stilldo://now")!
  static let sweep = URL(string: "stilldo://sweep")!
}

File: android/app/src/main/java/com/stilldo/widgets/WidgetStore.kt
package com.stilldo.widgets

import android.content.Context
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import org.json.JSONObject

/**
 * What the home-screen widgets are allowed to know.
 *
 * Keep in step with `src/widgets/snapshot.ts`. A widget provider runs in the
 * launcher's process, not the app's, so it cannot open the app's SQLite file —
 * it reads this summary out of shared preferences instead.
 */
data class WidgetTask(val id: Int, val title: String, val note: String)

data class WidgetSnapshot(
    val rightNow: WidgetTask?,
    val openCount: Int,
    private val caughtCount: Int,
    private val caughtDay: String,
    val sweepTime: String,
    val appearance: String,
) {
  /**
   * The tally is about the day it was written on. A widget outlives that, so
   * after midnight it reports nothing caught rather than yesterday's count.
   */
  fun caughtToday(on: String): Int = if (caughtDay == on) caughtCount else 0

  companion object {
    const val SUPPORTED_VERSION = 1

    /** A widget added before the app has ever been opened. */
    val EMPTY = WidgetSnapshot(null, 0, 0, "", "21:00", "system")

    /**
     * Null for anything this build cannot read — nothing written yet, or a
     * payload from a version of the app whose shape it does not know.
     */
    fun parse(json: String?): WidgetSnapshot? {
      if (json.isNullOrEmpty()) return null
      return try {
        val o = JSONObject(json)
        if (o.optInt("version") != SUPPORTED_VERSION) return null
        val now = o.optJSONObject("rightNow")
        WidgetSnapshot(
            rightNow =
                now?.let {
                  WidgetTask(it.optInt("id"), it.optString("title"), it.optString("note"))
                },
            openCount = o.optInt("openCount"),
            caughtCount = o.optInt("caughtToday"),
            caughtDay = o.optString("caughtDay"),
            sweepTime = o.optString("sweepTime", "21:00"),
            appearance = o.optString("appearance", "system"),
        )
      } catch (e: Exception) {
        null
      }
    }
  }
}

object WidgetStore {
  private const val FILE = "stilldo.widgets"
  private const val KEY = "snapshot"

  /** 'YYYY-MM-DD' in local time — the day boundary the app counts by. */
  fun today(): String = SimpleDateFormat("yyyy-MM-dd", Locale.US).format(Date())

  fun write(context: Context, json: String) {
    context
        .getSharedPreferences(FILE, Context.MODE_PRIVATE)
        .edit()
        .putString(KEY, json)
        .apply()
  }

  fun read(context: Context): WidgetSnapshot =
      WidgetSnapshot.parse(
          context.getSharedPreferences(FILE, Context.MODE_PRIVATE).getString(KEY, null))
          ?: WidgetSnapshot.EMPTY
}
